overlord/hookstate/hooks.go

```diff
@@ -72,6 +76,141 @@ func SetupPreRefreshHook(st *state.State, snapName string) *state.Task {
 	return task
 }
 
+type gateAutoRefreshHookHandler struct {
+	context *Context
+	refreshAppAwareness bool
+}
+
+func (h *gateAutoRefreshHookHandler) Before() error {
+	st := h.context.State()
+	st.Lock()
+	defer st.Unlock()
+
+	tr := config.NewTransaction(st)
+	experimentalRefreshAppAwareness, err := features.Flag(tr, features.RefreshAppAwareness)
+	if err != nil && !config.IsNoOption(err) {
+		return err
+	}
+	if !experimentalRefreshAppAwareness {
+		return nil
+	}
+
+	h.refreshAppAwareness = true
+	if err := runinhibit.LockWithHint(h.context.InstanceName(), runinhibit.HintInhibitedGateRefresh); err != nil {
```

Comment:
afaiu we need to hold temporarily the snap lock when we manipulate the run inhibit lock, same for all the places here